File: invertBinaryTree.py
# problem: https://leetcode.com/problems/invert-binary-tree/submissions/
# Runtime: 44 ms, faster than 6.02% of Python3 online submissions for Invert Binary Tree.
# Memory Usage: 14.3 MB, less than 44.67% of Python3 online submissions for Invert Binary Tree.


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def invertTree(self, root: TreeNode) -> TreeNode:
        if not root:
            return root
        root.left, root.right = self.invertTree(
            root.right), self.invertTree(root.left)
        return root

File: invertBinaryTree.cpp
// problem: https://leetcode.com/problems/invert-binary-tree/description/
// Runtime
// 3 ms
// Beats
// 62.83%
// Memory
// 9.8 MB
// Beats
// 46.85%

#include <bits/stdc++.h>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  TreeNode *invertTree(TreeNode *root) {
    if (!root)
      return root;
    swap(root->right, root->left);
    invertTree(root->right);
    invertTree(root->left);
    return root;
  }
};
